osu.Game.Rulesets.Osu/Skinning/Smoke.cs

```diff
@@ -0,0 +1,384 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using osu.Framework.Allocation;
+using osu.Framework.Extensions.Color4Extensions;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Colour;
+using osu.Framework.Graphics.Primitives;
+using osu.Framework.Graphics.Rendering;
+using osu.Framework.Graphics.Rendering.Vertices;
+using osu.Framework.Graphics.Shaders;
+using osu.Framework.Graphics.Textures;
+using osu.Framework.Utils;
+using osu.Game.Rulesets.Osu.UI;
+using osuTK;
+using osuTK.Graphics;
+
+namespace osu.Game.Rulesets.Osu.Skinning
+{
+    public abstract class Smoke : Drawable, ITexturedShaderDrawable
+    {
+        public IShader? TextureShader { get; private set; }
+        public IShader? RoundedTextureShader { get; private set; }
+
+        private float? radius;
+
+        protected float Radius
+        {
+            get => radius ?? Texture?.DisplayWidth * 0.165f ?? 3;
+            set => radius = value;
+        }
+
+        protected Texture? Texture { get; set; }
+
+        protected double SmokeStartTime { get; private set; } = double.MinValue;
+
+        protected double SmokeEndTime { get; private set; } = double.MaxValue;
+
+        private Vector2 topLeft;
+
+        protected Vector2 TopLeft
+        {
+            get => topLeft;
+            set
+            {
+                if (topLeft == value)
+                    return;
+
+                topLeft = value;
+                Invalidate(Invalidation.Layout);
+            }
+        }
+
+        private Vector2 bottomRight;
+
+        protected Vector2 BottomRight
+        {
+            get => bottomRight;
+            set
+            {
+                if (bottomRight == value)
+                    return;
+
+                bottomRight = value;
+                Invalidate(Invalidation.Layout);
+            }
+        }
+
+        protected abstract double LifetimeAfterSmokeEnd { get; }
+        protected virtual float PointInterval => Radius * 7f / 8;
+        protected bool IsActive { get; private set; }
+
+        protected readonly List<SmokePoint> SmokePoints = new List<SmokePoint>();
+
+        private float totalDistance;
+        private Vector2? lastPosition;
+
+        private const int max_point_count = 18_000;
```

Comment:
I've reduced this from 72000 because that's a number of quads that's reached well after my FPS dips below 30.